File: TSP/Java/src/tsp/Arc.java
package tsp;

class Arc implements Comparable<Arc>
{
    private int source;
    private int target;
    private double distance;
    
    public double getDistance()
    {
        return distance;
    }

    public Arc(int source, int target, double [][] dM)
    {
        this.source = source;
        this.target = target;
        distance = dM[source][target];        
    }

    public int getSource()
    {
        return source;
    }

    public int getTarget()
    {
        return target;
    }

    public String toString()
    {
        return "(" + source + "," + target + ")";
    }

    private int normalize(double lol)
    {
        return (int) (lol / Math.abs(lol));
    }
    
    public int compareTo(Arc other)
    {
        return normalize(this.getDistance() - other.getDistance());
    }
}

File: TSP/Cplusplus/Arc.cpp
#include "./DecreasingArcHeuristicTSP.hpp"

DecreasingArcHeuristicTSP::Solver::Arc::Arc(int i, int j, const double **distance_)
    :source(i), target(j), distance(distance_[i][j])
{
}

bool DecreasingArcHeuristicTSP::Solver::Arc::compare(Arc *a, Arc *b)
{
    return a->distance < b->distance;
}

double DecreasingArcHeuristicTSP::Solver::Arc::getDistance()
{
    return distance;
}

int DecreasingArcHeuristicTSP::Solver::Arc::getSource()
{
    return source;
}

int DecreasingArcHeuristicTSP::Solver::Arc::getTarget()
{
    return target;
}

DecreasingArcHeuristicTSP::Solver::Arc::operator string()
{
    return "(" + to_string(source) + "," + to_string(target) + ")";
}
